### PR TITLE
Add support for parameterized SipHash

### PR DESCRIPTION
**Pull request by tmshort**
The core SipHash supports either 8 or 16-byte output and a configurable
number of rounds.
The default behavior, as added to EVP, is to use 16-byte output and
2,4 rounds, which matches the behavior of most implementations.

Note that documentation has not been updated, as there was some doc updates
for Poly1305 that I did in PR #2128 that I want to build upon.

##### Checklist

- [x] documentation is added or updated
- [X] tests are added or updated
- [X] CLA is signed

### CODE REVIEW

File: test/testutil.c [hunk 1]
```diff
@@ -161,3 +161,13 @@ int strings_equal(const char *desc, const char *s1, const char *s2)
...
+
+/* TODO : hex decoder should be implemented */
+void hexdump(const unsigned char *a, size_t len)
+{
+    size_t i;
+
+    for (i = 0; i < len; i++)
+        fprintf(stderr, "%02x", a[i]);
+}
+
```
Comment (levitte) [inline test/testutil.c]: This could really be a separate PR

Comment (tmshort) [inline test/testutil.c]: Yeah; I pulled it out of poly1305_internal_test.c

Comment (richsalz) [inline test/testutil.c]: Look at BIO_hex_string.

Comment (richsalz) [inline test/testutil.c]: Cleaning up all "hexdump" to use that one common routine is a separate PR, but let's not make it the situation worse with this one.

Comment (tmshort) [inline test/testutil.c]: Not necessary with BIO_hex_string, but I undid the changes to poly1305.

**Comment by levitte** [comment]
This needs a rebase

**Comment by tmshort** [comment]
Yup, with the Poly1305 commit.

**Comment by tmshort** [comment]
Rebased.

**Comment by levitte** [comment]
Looks good to me.

File: crypto/evp/p_lib.c [hunk 2]
```diff
@@ -263,6 +263,20 @@ const unsigned char *EVP_PKEY_get0_poly1305(const EVP_PKEY *pkey, size_t *len)
 }
 #endif
 
+#ifndef OPENSSL_NO_SIPHASH
+const unsigned char *EVP_PKEY_get0_siphash(const EVP_PKEY *pkey, size_t *len)
+{
+    ASN1_OCTET_STRING *os = NULL;
+    if (pkey->type != EVP_PKEY_SIPHASH) {
```
Comment (richsalz) [inline crypto/evp/p_lib.c]: blank line after decl.

File: crypto/include/internal/siphash.h [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+ */
+
+#include <stddef.h>
+
+#define SIPHASH_BLOCK_SIZE        8
+#define SIPHASH_KEY_SIZE         16
+#define SIPHASH_MIN_DIGEST_SIZE   8
+#define SIPHASH_MAX_DIGEST_SIZE  16
+
+typedef struct siphash_context SIPHASH;
```
Comment (richsalz) [inline crypto/include/internal/siphash.h]: we don't have consistency here, but maybe siphash_st is the more commonly used style?

File: crypto/siphash/siphash_ameth.c [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+static int siphash_size(const EVP_PKEY *pkey)
+{
+    return SIPHASH_MAX_DIGEST_SIZE;
+}
+
+static void siphash_key_free(EVP_PKEY *pkey)
+{
+    ASN1_OCTET_STRING *os = EVP_PKEY_get0(pkey);
+    if (os != NULL) {
```
Comment (richsalz) [inline crypto/siphash/siphash_ameth.c]: blank line after decls.

File: crypto/siphash/siphash_ameth.c [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        if (os->data != NULL)
+            OPENSSL_cleanse(os->data, os->length);
+        ASN1_OCTET_STRING_free(os);
+    }
+}
+
+static int siphash_pkey_ctrl(EVP_PKEY *pkey, int op, long arg1, void *arg2)
+{
+    /* nothing, (including ASN1_PKEY_CTRL_DEFAULT_MD_NID), is supported */
```
Comment (richsalz) [inline crypto/siphash/siphash_ameth.c]: remove, the, first, comma.
or (remove the parens)

File: crypto/siphash/siphash_pmeth.c [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+#include <openssl/x509.h>
+#include <openssl/x509v3.h>
+#include <openssl/evp.h>
+#include "internal/siphash.h"
+#include "siphash_local.h"
+#include "internal/evp_int.h"
+
+/* SIPHASH pkey context structure */
+
+typedef struct {
```
Comment (richsalz) [inline crypto/siphash/siphash_pmeth.c]: put a name here, ```siphash_pkey_ctx_st```

File: doc/man3/EVP_PKEY_CTX_ctrl.pod [hunk 7]
```diff
@@ -49,7 +49,7 @@ The control command is indicated in B<cmd> and any additional arguments in
 B<p1> and B<p2>.
 
 For B<cmd> = B<EVP_PKEY_CTRL_SET_MAC_KEY>, B<p1> is the length of the MAC key,
-and B<p2> is MAC key. This is used by Poly1305, HMAC and CMAC.
+and B<p2> is MAC key. This is used by Poly1305, SipHash, HMAC and CMAC.
```
Comment (richsalz) [inline doc/man3/EVP_PKEY_CTX_ctrl.pod]: all that work and this is the only doc update needed? :)  the joy of standard API's.

File: crypto/siphash/siphash_ameth.c (hunk 5 above)
Comment (tmshort) [inline crypto/siphash/siphash_ameth.c]: Three commas in a row? oh my,

File: INSTALL [hunk 8]
```diff
@@ -494,8 +494,9 @@
                    Build without support for the specified algorithm, where
                    <alg> is one of: bf, blake2, camellia, cast, chacha, cmac,
                    des, dh, dsa, ecdh, ecdsa, idea, md4, mdc2, ocb, poly1305,
-                   rc2, rc4, rmd160, scrypt, seed or whirlpool. The "ripemd"
-                   algorithm is deprecated and if used is synonymous with rmd160.
+                   rc2, rc4, rmd160, scrypt, seed, siphash or whirlpool. The
+		   "ripemd" algorithm is deprecated and if used is synonymous
```
Comment (tmshort) [inline INSTALL]: might be some white-space issues here.

File: doc/man3/EVP_PKEY_CTX_ctrl.pod (hunk 7 above)
Comment (tmshort) [inline doc/man3/EVP_PKEY_CTX_ctrl.pod]: I don't believe the EVP_PKEY_SIPHASH (or other similar ones) have a place for documentation. But I should check.

Comment (richsalz) [inline doc/man3/EVP_PKEY_CTX_ctrl.pod]: it was a comment on how good standard API's are.  not asking for more work on this.

File: INSTALL (hunk 8 above)
Comment (richsalz) [inline INSTALL]: check space/TAB issues.  GH sometimes gets confused, too.

File: doc/man3/EVP_PKEY_CTX_ctrl.pod (hunk 7 above)
Comment (tmshort) [inline doc/man3/EVP_PKEY_CTX_ctrl.pod]: Oh, I know! :) But I want to be thorough.

**Comment by tmshort** [comment]
There already is a no-siphash option. It's currently on by default

File: crypto/siphash/siphash_pmeth.c (hunk 6 above)
Comment (levitte) [inline crypto/siphash/siphash_pmeth.c]: Uhmmmm...  why is that important?

Comment (richsalz) [inline crypto/siphash/siphash_pmeth.c]: it's not important.

**Comment by tmshort** [comment]
Rebased, comments resolved.

File: doc/man3/EVP_PKEY_CTX_ctrl.pod (hunk 7 above)
Comment (tmshort) [inline doc/man3/EVP_PKEY_CTX_ctrl.pod]: Nothing referencing stuff like EVP_PKEY_HMAC, etc.

**Comment by richsalz** [comment]
@levitte please reconfirm and i will merge.

**Comment by richsalz** [comment]
3f5616d on master.  Thanks!
